### PR TITLE
test(github): Snapshot all HTTP requests in test

### PR DESCRIPTION
**Pull request by zharinov**
After merge, we can confidently switch to new `util/http` wrapper.

Ref: #5865

### CODE REVIEW

**Comment by viceice** [comment]
I'm thinking of making all webrequest tests with `nock`, so we are not dependend on any http layer other than the default node request api. We can then easily change any http layer between

**Comment by zharinov** [comment]
Should be doable, with snapshots obtained via [call recording](https://github.com/nock/nock#output_objects-option) feature.

**Comment by zharinov** [comment]
Do you want to see it in this PR?

**Comment by viceice** [comment]
Sure, if you like to do this. 🙃

**Comment by zharinov** [comment]
`Nock` greatly increases tests readability here. But I'm having strange behavior running multiple tests: only the first one is passing, but other are okay when running separately. It returns previous responses for particular URL even after re-define, using `once()` and so on...
image

**Comment by viceice** [comment]
You need to remember the retries of got on failures. You should also reset all mock before each test and check is done on mock after the test.

**Comment by zharinov** [comment]
I've made sure everything resets after each `it(...)` block, but the only working solution is to run them in separate `describe(...)` sections. Running Jest `--runInBand` doesn't help. Minimal node script reproducing the issue works fine.
So weird.

**Comment by viceice** [comment]
@zharinov if you push the current state i can have a look at the issue. I've fighted the same issue yesterday but with got v11 (silently fails while retriying)

**Comment by zharinov** [comment]
` yarn jest -t 'getRepoForceRebase' lib/platform/github/index.spec.ts`

**Comment by viceice** [comment]
Again caching is destroying your tests

**Comment by viceice** [comment]
@zharinov Works 🎉 

https://github.com/renovatebot/renovate/blob/9cf87d418bb33cd9daed268b19c8ce78a4888985/lib/platform/github/index.spec.ts#L63

**Comment by zharinov** [comment]
OMG, thanks!

**Comment by viceice** [comment]
we have a got get cache 

https://github.com/renovatebot/renovate/blob/bb1ae6b7c131fc537466acc6598d67cdab9d7a5e/lib/util/got/cache-get.ts#L21-L39

**Comment by zharinov** [comment]
Playing around `nock.isDone()` and `nock.pendingMocks()`, I've found lots of missing mocks and pending requests in the tests (e.g. because of `try-catch` blocks). Now I'm about to declare them explicitly. When I finish this PR, `util/http/mock.ts` will help to handle it for other http-related tests.

**Comment by rarkins** [comment]
@zharinov FYI I'm refactoring out the global `repoCache` in #6098 right now, in case that helps. I had hoped we could have an option to disable cache for all test but then realized that quite a few of them don't work without it.

**Comment by rarkins** [comment]
Non-global cache now removed and merged into this branch.

File: lib/platform/github/index.spec.ts [hunk 1]
```diff
@@ -65,845 +69,752 @@ describe('platform/github', () => {
...
+          ],
+        });
+      await github.initRepo({
+        repository: 'some/repo',
+      } as any);
       const res = await github.getBranchStatus('somebranch', []);
       expect(res).toEqual(BranchStatus.red);
+      expect(httpMock.getTrace()).toMatchSnapshot();
     });
     it('should suceed if no status and all passed check runs', async () => {
```
Comment (JamieMagee) [inline lib/platform/github/index.spec.ts:622]:
```suggestion
    it('should succeed if no status and all passed check runs', async () => {
```

**Review by JamieMagee — DISMISSED** [review]
Amazing. Way more readable.

Are you still planning on replacing all the calls in this PR? Or separate them out?

**Comment by zharinov** [comment]
I planned to merge it only for `platform/github`, in order to confidently refactor its' internal `got` calls. So this PR is ready for review/merge I believe.

**Pull request merged**